File: sdk/java/picodroid/widget/LinearLayout.java
// SPDX-License-Identifier: GPL-3.0-only
package picodroid.widget;

import picodroid.content.Context;
import picodroid.view.ViewGroup;

public class LinearLayout extends ViewGroup {
  public static final int HORIZONTAL = 0;
  public static final int VERTICAL = 1;

  private int orientation;

  public LinearLayout() {
    super(nativeCreate());
    this.orientation = VERTICAL;
  }

  public LinearLayout(Context ctx) {
    super(nativeCreate());
    this.orientation = VERTICAL;
  }

  private static native int nativeCreate();

  public native void setOrientation(int orientation);

  /** Gap in pixels between adjacent children. Default 0. */
  public native void setSpacing(int px);

  /**
   * Set the alignment of the children. Mirrors {@code android.widget.LinearLayout#setGravity(int)}:
   * pass {@link picodroid.view.Gravity} constants, which may name both axes ({@code Gravity.BOTTOM
   * | Gravity.RIGHT}). The horizontal bits place the children of a {@link #HORIZONTAL} layout along
   * its length and a {@link #VERTICAL} one across it, and the vertical bits the other way round, as
   * on Android; call it after {@link #setOrientation}, which is what decides which is which.
   *
   * <p>Two divergences: an axis the gravity does not name keeps centring rather than falling back
   * to the start, and {@code FILL} places at the start instead of stretching the child. Per-child
   * {@code LayoutParams.gravity} is not yet applied — that is part of the LayoutParams milestone.
   */
  public native void setGravity(int gravity);

  /**
   * Mirrors {@code android.widget.LinearLayout.LayoutParams}. Adds {@code weight} (mapped to LVGL
   * {@code lv_obj_set_flex_grow}) and {@code gravity} (per-child alignment along the cross axis).
   */
  public static class LayoutParams extends ViewGroup.LayoutParams {
    public float weight;
    public int gravity;

    public LayoutParams(int width, int height) {
      super(width, height);
    }

    public LayoutParams(int width, int height, float weight) {
      super(width, height);
      this.weight = weight;
    }
  }
}

File: crates/picodroid-core/src/graphics/lvgl/widgets/linear_layout.rs
// SPDX-License-Identifier: GPL-3.0-only
//! LVGL impl of `LinearLayout` (LVGL `lv_obj` with flex flow).

use crate::lvgl_ffi::*;

use super::super::handle_table;
use super::super::lifecycle;
use super::super::style_batch;

pub(in crate::graphics) fn create() -> i32 {
    let ptr = unsafe {
        let o = lv_obj_create(lifecycle::screen_ptr());
        // Android LinearLayout never scrolls — use ScrollView for that. Clearing
        // SCROLLABLE also kills the stray scrollbars LVGL would otherwise draw
        // when content brushes the inner edge.
        lv_obj_remove_flag(o, LV_OBJ_FLAG_SCROLLABLE);
        // Flat first — no theme border, fill, corners or padding, as on Android
        // (`frame_layout::make_flat`) — since the strip removes local styles
        // too, and the flex flow is one. The strip and the two flex sets share
        // one refresh (style_batch.rs); the six padding writes the strip
        // replaced were most of this native's 1.7 ms on the RP2350.
        style_batch::with_one_refresh(o, LV_STYLE_PAD_TOP, || {
            super::frame_layout::make_flat(o);
            lv_obj_set_flex_flow(o, LV_FLEX_FLOW_COLUMN);
            lv_obj_set_flex_align(
                o,
                LV_FLEX_ALIGN_START,
                LV_FLEX_ALIGN_CENTER,
                LV_FLEX_ALIGN_CENTER,
            );
        });
        o
    };
    handle_table::register(ptr)
}

/// `orientation`: 0 = horizontal (row), non-zero = vertical (column).
pub(in crate::graphics) fn set_orientation(id: i32, orientation: i32) {
    let flow = if orientation == 0 {
        LV_FLEX_FLOW_ROW
    } else {
        LV_FLEX_FLOW_COLUMN
    };
    unsafe { lv_obj_set_flex_flow(handle_table::lookup(id), flow) };
}

/// Gap in pixels between adjacent children. Sets both row and column gap so
/// the value applies whether the layout is later switched to horizontal flow.
pub(in crate::graphics) fn set_spacing(id: i32, spacing: i32) {
    unsafe {
        let o = handle_table::lookup(id);
        lv_obj_set_style_pad_row(o, spacing, 0);
        lv_obj_set_style_pad_column(o, spacing, 0);
    }
}

/// Place the children from an Android gravity bitmask.
///
/// Which of Android's two axes is this layout's main one follows the flow
/// `setOrientation` wrote, which LVGL keeps as a style property — so `RIGHT`
/// aligns a row's children at its right edge and centres a column's children
/// horizontally, exactly as `android.widget.LinearLayout.setGravity` does.
/// `gravity::flex_align` holds the decoding, and its divergences, in one
/// tested place.
pub(in crate::graphics) fn set_gravity(id: i32, gravity: i32) {
    let obj = handle_table::lookup(id);
    if obj.is_null() {
        return; // stale handle — mutating a destroyed View is a no-op
    }
    unsafe {
        let flow = lv_obj_get_style_prop(obj, LV_PART_MAIN, LV_STYLE_FLEX_FLOW).num as u32;
        // Only ROW and COLUMN are ever set, and COLUMN is the low bit of every
        // column flow LVGL defines.
        let (main, cross) = super::gravity::flex_align(gravity, flow & LV_FLEX_FLOW_COLUMN != 0);
        lv_obj_set_flex_align(obj, main, cross, cross);
    }
}
